flink-runtime/src/main/java/org/apache/flink/runtime/metrics/dump/MetricDumpSerialization.java

```diff
@@ -191,62 +233,65 @@ private static void serializeMeter(DataOutputStream dos, Meter meter) throws IOE
 		 *
 		 * @param data serialized metrics
 		 * @return A list containing the deserialized metrics.
-		 * @throws IOException
 		 */
-		public List<MetricDump> deserialize(byte[] data) throws IOException {
-			ByteArrayInputStream bais = new ByteArrayInputStream(data);
-			DataInputStream dis = new DataInputStream(bais);
 
-			int numCounters = dis.readInt();
-			int numGauges = dis.readInt();
-			int numHistograms = dis.readInt();
-			int numMeters = dis.readInt();
+		public List<MetricDump> deserialize(MetricDumpSerialization.MetricSerializationResult data) {
+			DataInputView in = new DataInputDeserializer(data.data, 0, data.data.length);
 
-			List<MetricDump> metrics = new ArrayList<>(numCounters + numGauges + numHistograms);
+			List<MetricDump> metrics = new ArrayList<>(data.numCounters + data.numGauges + data.numHistograms + data.numMeters);
 
-			for (int x = 0; x < numCounters; x++) {
-				metrics.add(deserializeCounter(dis));
+			for (int x = 0; x < data.numCounters; x++) {
+				try {
+					metrics.add(deserializeCounter(in));
+				} catch (Exception e) {
+					LOG.warn("Failed to deserialize counter.", e);
+				}
 			}
 
-			for (int x = 0; x < numGauges; x++) {
-				metrics.add(deserializeGauge(dis));
+			for (int x = 0; x < data.numGauges; x++) {
+				try {
+					metrics.add(deserializeGauge(in));
+				} catch (Exception e) {
+					LOG.warn("Failed to deserialize counter.", e);
+				}
 			}
 
-			for (int x = 0; x < numHistograms; x++) {
-				metrics.add(deserializeHistogram(dis));
+			for (int x = 0; x < data.numHistograms; x++) {
+				try {
+					metrics.add(deserializeHistogram(in));
+				} catch (Exception e) {
+					LOG.warn("Failed to deserialize counter.", e);
+				}
 			}
 
-			for (int x = 0; x < numMeters; x++) {
-				metrics.add(deserializeMeter(dis));
+			for (int x = 0; x < data.numMeters; x++) {
+				try {
+					metrics.add(deserializeMeter(in));
+				} catch (Exception e) {
+					LOG.warn("Failed to deserialize counter.", e);
+				}
 			}
-
-			return metrics;
 		}
 	}
 
-	private static String deserializeString(DataInputStream dis) throws IOException {
-		int stringLength = dis.readInt();
-		byte[] bytes = new byte[stringLength];
-		dis.readFully(bytes);
-		return new String(bytes);
-	}
 
-	private static MetricDump.CounterDump deserializeCounter(DataInputStream dis) throws IOException {
+	private static MetricDump.CounterDump deserializeCounter(DataInputView dis) throws IOException {
 		QueryScopeInfo scope = deserializeMetricInfo(dis);
-		String name = deserializeString(dis);
-		return new MetricDump.CounterDump(scope, name, dis.readLong());
+		String name = dis.readUTF();
+		long count = dis.readLong();
+		return new MetricDump.CounterDump(scope, name, count);
```

Comment:
Personally for short methods i think it's overkill. I would do it for methods like `deserializaHistogram` though.